packages/Chem/src/chem-searches.ts

```diff
@@ -82,116 +86,151 @@ function _chemGetDiversities(limit: number, molStringsColumn: DG.Column, fingerp
     lastColumnInvalidated = invalidatedColumnKey(molCol);
     molCol.setTag(FING_COL_TAGS.invalidatedForVersion, String(molCol.version + 1));
   }
 }
 
-function checkForFingerprintsColumn(col: DG.Column, fingerprintsType: Fingerprint): Uint8Array[] | null {
-  const colNameTag = '.' + fingerprintsType + '.Column';
-  const colVerTag = '.' + fingerprintsType + '.Version';
+
+function checkForSavedColumn(col: DG.Column, colTag: Fingerprint | string): DG.Column<any> | null{
+  const colNameTag = '.' + colTag + '.Column';
+  const colVerTag = '.' + colTag + '.Version';
 
   if (col.getTag(colNameTag) &&
       col.getTag(colVerTag) == String(col.version) &&
       col.dataFrame) {
-    const fingCol = col.dataFrame.columns.byName(col.getTag(colNameTag));
-    if (fingCol)
-      return fingCol.toList();
+    const savedCol = col.dataFrame.columns.byName(col.getTag(colNameTag));
+    return savedCol;
   }
   return null;
 }
 
-function saveFingerprintsToCol(col: DG.Column, fgs: Uint8Array[],
-  fingerprintsType: Fingerprint, createdMols: boolean): void {
+function saveColumns(col: DG.Column, data: ((Uint8Array | null)[] | (string | null)[])[],
+  tags: string[], colTypes: DG.COLUMN_TYPE[], createdMols: boolean): void {
+  
+  for (let i = 0; i < data.length; i++)
+    saveColumn(col, data[i], tags[i], colTypes[i]);
+
+  const colVersion = createdMols ? col.version + data.length + 2 : col.version + data.length + 1;
+  if (createdMols)
+    col.setTag(FING_COL_TAGS.molsCreatedForVersion, String(colVersion));
+  for (let i = 0; i < data.length; i++)
+    col.setTag('.' + tags[i] + '.Version', String(colVersion));
+  col.setTag(FING_COL_TAGS.invalidatedForVersion, String(colVersion));
+}
+
+
+function saveColumn(col: DG.Column, data: (Uint8Array | null)[] | (string | null)[],
+  tagName: string, colType: DG.COLUMN_TYPE): void {
   if (!col.dataFrame)
     throw new Error('Column has no parent dataframe');
 
-  const colNameTag = '.' + fingerprintsType + '.Column';
-  const colVerTag = '.' + fingerprintsType + '.Version';
+  const colNameTag = '.' + tagName + '.Column';
 
-  const fingerprintColumnName = col.getTag(colNameTag) ??
-    '~' + col.name + fingerprintsType + 'Fingerprints';
+  const savedColumnName = col.getTag(colNameTag) ??
+    '~' + col.name + '.' + tagName;
   const df = col.dataFrame;
   const newCol: DG.Column<Uint8Array> = df.columns.getOrCreate(
-    fingerprintColumnName, DG.COLUMN_TYPE.BYTE_ARRAY, fgs.length);
+    savedColumnName, colType, data.length);
 
-  newCol.init((i) => fgs[i]);
+  newCol.init((i) => data[i]);
 
-  col.setTag(colNameTag, fingerprintColumnName);
-  if (createdMols)
-    col.setTag(FING_COL_TAGS.molsCreatedForVersion, String(col.version + 3));
-  col.setTag(colVerTag, String(col.version + 2));
-  col.setTag(FING_COL_TAGS.invalidatedForVersion, String(col.version + 1));
+  col.setTag(colNameTag, savedColumnName);
+}
+
+async function invalidateAndSaveColumns(molCol: DG.Column, fingerprintsType: Fingerprint,
```

Comment:
The "returnSmiles" looks like a code smell to me, can we make it simpler?

Reply:
Not sure if we can get rid of this additional parameter. The situation is the following: when we calculate pattern fp we also might need to return canonical smiles (and I want to do this in the same function not to create mol objects repeatedly). But we also might not need to calculate canonical smiles at least in 2 cases:
- we are creating Morgan fp;
- we already have a column which contains canonical smiles (for instance when I extracted molecules from Chembl I aleady had a column with canonical smiles and did not want to spend time and memory to create additional column).

So I need this additional parameter to know if creating of canonical smiles is required. @skalkin , do you think there is a better way to do this?